diff --git a/core/lib/Drupal/Core/Asset/LibraryDependencyResolver.php b/core/lib/Drupal/Core/Asset/LibraryDependencyResolver.php
index 4c31c1b1..5f13561f 100644
--- a/core/lib/Drupal/Core/Asset/LibraryDependencyResolver.php
+++ b/core/lib/Drupal/Core/Asset/LibraryDependencyResolver.php
@@ -64,7 +64,7 @@ public function getLibrariesWithDependencies(array $libraries) {
   protected function doGetDependencies(array $libraries_with_unresolved_dependencies, array $final_libraries = []) {
     foreach ($libraries_with_unresolved_dependencies as $library) {
       if (!isset($final_libraries[$library])) {
-        [$extension, $name] = explode('/', $library, 2);
+        [$extension, $name] = array_pad(explode('/', $library), 2, NULL);
         $definition = $this->libraryDiscovery->getLibraryByName($extension, $name);
         if (!empty($definition['dependencies'])) {
           $final_libraries = $this->doGetDependencies($definition['dependencies'], $final_libraries);
